### PR TITLE
Inject request into template context

### PR DESCRIPTION
**Pull request by jtraub**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

This PR introduces implicit request injection into template context. Injected request will override existing `request` value that user might have passed into context.

### CODE REVIEW

File: starlite/datastructures/response_containers.py [hunk 1]
```diff
@@ -59,7 +60,12 @@ class Config(BaseConfig):
 
     @abstractmethod
```
Comment (Goldziher) [inline starlite/datastructures/response_containers.py:61]: why do you need to update the abstract method with the `request` kwarg? you can define it only for `Template` IMO.

File: starlite/datastructures/response_containers.py [hunk 2]
```diff
@@ -213,11 +240,12 @@ def to_response(
         from starlite.exceptions import ImproperlyConfiguredException
         from starlite.response import TemplateResponse
 
+        context = self.context or {}
         if not app.template_engine:
             raise ImproperlyConfiguredException("Template engine is not configured")
         return TemplateResponse(
             background=self.background,
-            context=self.context,
+            context={**context, "request": request},
```
Comment (Goldziher) [inline starlite/datastructures/response_containers.py:248]: you dont need to hardcode it here in a way that will overide the user. I would suffest doing this:

```python
context = self.context or {}
context.setdefault(request=request)
```

File: starlite/response.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def __init__(
         template_name: str,
         template_engine: "TemplateEngineProtocol",
         status_code: int,
-        context: Optional[Dict[str, Any]] = None,
+        context: Dict[str, Any],
```
Comment (Goldziher) [inline starlite/response.py:111]: why is this required?

File: starlite/datastructures/response_containers.py (hunk 1 above)
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:61]: ok, i see why you need this value.

File: docs/usage/15-templating.md [hunk 4]
```diff
@@ -50,6 +50,9 @@ The `name` kwarg passed to the `Template` class is the filename for the given te
 directories specifies for this file until it finds it or an exception will be raised. The `context` kwarg is a
 dictionary specifying context data that is passed to the engine.
 
+Starlite automatically injects current request into context under `request` name. Note that if you try pass any value
```
Comment (Goldziher) [inline docs/usage/15-templating.md:53]: I dont think this is the way to go, instead:

```text
Starlite injects the current Request instance into the template context under the `request` key by default. 
```

**Review by Goldziher — COMMENTED** [review]
Nice work! Some small adjustments and we are good to go.

File: starlite/datastructures/response_containers.py (hunk 2 above)
Comment (jtraub) [reply starlite/datastructures/response_containers.py:248]: Uhm, as a user I would prefer to be sure that some template globals won't be overridden accidentally allowing me to simplify jinja2/make function code and some of my tests.

File: starlite/response.py (hunk 3 above)
Comment (jtraub) [reply starlite/response.py:111]: At the point context is guaranteed to be a dict with `request` so removing `Optional` is just making typing a bit more strict.

File: tests/handlers/http/test_to_response.py [hunk 5]
```diff
@@ -120,7 +120,7 @@ def test_function() -> Response:
     with create_test_client(test_function) as client:
         http_route: HTTPRoute = client.app.routes[0]  # type: ignore
         route_handler = http_route.route_handlers[0]
-        response = await route_handler.to_response(data=route_handler.fn(), plugins=[], app=None)  # type: ignore
+        response = await route_handler.to_response(data=route_handler.fn(), plugins=[], app=None, request=None)  # type: ignore
```
[the remark this answers is not in the document]
Comment (jtraub) [reply tests/handlers/http/test_to_response.py:123]: @Goldziher I wanted to keep test_to_response.py simple - hence passing `None` instead of `request` as we need to test that `handler.to_response` returns proper response and request doesn't really matter here.

[the remark this answers is not in the document]
Comment (Goldziher) [reply tests/handlers/http/test_to_response.py:123]: sure but it will save you on a lot of type ignores. just make one request instance nad use it - will actually save you code.

File: starlite/datastructures/response_containers.py (hunk 2 above)
Comment (Goldziher) [reply starlite/datastructures/response_containers.py:248]: ok.

File: tests/handlers/http/test_to_response.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jtraub) [reply tests/handlers/http/test_to_response.py:123]: It is 4AM for me so I will update the code tomorrow in the morning.

I should note though that `# type: ignore` comments won't go away because of `app=None` and `data=route_handler.fn()` arguments

File: starlite/datastructures/response_containers.py (hunk 2 above)
Comment (jtraub) [reply starlite/datastructures/response_containers.py:248]: @Goldziher sorry, I am a bit tired and didn't understand your last comment.

Do you want me to leave it as is (forbidding the user to override `request` global in templates) or remove overriding? If you want me to change the behavior I will do it tomorrow.

Comment (Goldziher) [reply starlite/datastructures/response_containers.py:248]: Ok. Let's go with your preference
